icu4c/source/i18n/collationruleparser.cpp

```diff
@@ -613,18 +613,24 @@ CollationRuleParser::parseSetting(UErrorCode &errorCode) {
                 return;
             }
             // localeID minus all keywords
-            char baseID[ULOC_FULLNAME_CAPACITY];
-            int32_t length = uloc_getBaseName(localeID.data(), baseID, ULOC_FULLNAME_CAPACITY, &errorCode);
-            if(U_FAILURE(errorCode) || length >= ULOC_KEYWORDS_CAPACITY) {
+            CharString baseID = ulocimp_getBaseName(localeID.toStringPiece(), errorCode);
+            if (U_FAILURE(errorCode)) {
                 errorCode = U_ZERO_ERROR;
                 setParseError("expected language tag in [import langTag]", errorCode);
                 return;
             }
-            if(length == 0) {
-                uprv_strcpy(baseID, "root");
-            } else if(*baseID == '_') {
-                uprv_memmove(baseID + 3, baseID, length + 1);
-                uprv_memcpy(baseID, "und", 3);
+            if (baseID.isEmpty()) {
+                baseID.copyFrom("root", errorCode);
+            } else if (baseID[0] == '_') {
+                // CharString doesn't have any insert() method, only append().
```

Comment:
FYI: We could add a CharString.prepend(StringPiece, UErrorCode) method

Reply:
Let's keep that in mind if this functionality is found to be needed somewhere else too.